src/main/kotlin/ecommerce/service/CartService.kt

```diff
@@ -0,0 +1,78 @@
+package ecommerce.service
+
+import ecommerce.dto.CartItemRequest
+import ecommerce.dto.CartItemResponse
+import ecommerce.exception.NotFoundException
+import ecommerce.model.Cart
+import ecommerce.model.CartItem
+import ecommerce.model.mapper.CartItemMapper
+import ecommerce.repository.CartRepository
+import ecommerce.repository.MemberRepository
+import ecommerce.repository.ProductRepository
+import org.springframework.data.domain.PageImpl
+import org.springframework.data.domain.PageRequest
+import org.springframework.data.domain.Sort
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import kotlin.math.min
+
+@Transactional
+@Service
+class CartService(
+    private val cartRepository: CartRepository,
+    private val productRepository: ProductRepository,
+    private val memberRepository: MemberRepository,
+) {
+    fun findCart(memberId: Long): Cart {
+        return cartRepository.findCartByMemberId(memberId)
+            ?: throw RuntimeException("not found something ... ")
+    }
+
+    fun addItem(
+        memberId: Long,
+        request: CartItemRequest,
+    ): CartItem {
+        val member = memberRepository.findById(memberId).orElseThrow { NotFoundException() }
+        val product =
+            productRepository.findById(request.productId).orElseThrow { NotFoundException() }
+
+        val cart =
+            cartRepository.findCartByMemberId(memberId)
+                ?: cartRepository.save(Cart(member))
+
+        val item = cart.addItem(product, request.quantity)
+        cartRepository.save(cart)
+        return item
+    }
+
+    fun deleteItem(
+        memberId: Long,
+        request: CartItemRequest,
+    ) {
+        val member = memberRepository.findById(memberId).orElseThrow { NotFoundException() }
+        val product =
+            productRepository.findById(request.productId).orElseThrow { NotFoundException() }
+
+        val cart =
+            cartRepository.findCartByMemberId(memberId)
+                ?: cartRepository.save(Cart(member))
+
+        cart.removeItem(product, request.quantity)
+        cartRepository.save(cart)
+    }
+
+    fun getPages(
+        memberId: Long,
+        page: Int,
+        size: Int,
+    ): PageImpl<CartItemResponse> {
+        val cart = findCart(memberId)
+        val itemResponses = cart.items.map { CartItemMapper.toResponse(it) }
+        val pageRequest = PageRequest.of(page, size, Sort.by("productName"))
+        val start = pageRequest.offset.toInt()
+        val end = min(start + pageRequest.pageSize, itemResponses.size)
+
+        val pageContent = itemResponses.subList(start, end)
+        return PageImpl<CartItemResponse>(pageContent, pageRequest, itemResponses.size.toLong())
+    }
```

Comment:
You did well, but it feels a bit complex.

How about this?

```suggestion
    ): Page<CartItemResponse> {
        val cart = findCart(memberId)
        val itemResponses = cart.items.map { CartItemMapper.toResponse(it) }
        val pageRequest = PageRequest.of(page, size, Sort.by("productName"))

        // You need to implement `toResponse()` though.
        return cartRepository.findAll(pageRequest).toResponse()
    }
```

Reply:
The actual implementation was handled a bit differently.  

Since I wanted to perform pagination directly at the database, I decided to use `CartItemRepository` instead of `CartRepository`.

```kotlin
 ): Page<CartItemResponse>  {
        val pageable = PageRequest.of(page, size, Sort.by("product.name"))
        return cartItemRepository
            .findByCartMemberId(memberId, pageable)
            .map(CartItemMapper::toResponse)
    }
```

Reply:
applied [here](https://github.com/herotechcourse/spring-ecommerce-order/pull/15/commits/dd8d9e4da000d5d8ae832f2d115d8f346e187472)